_source/_posts/2022-02-22-manage-dotnet-microservices.md

```diff
@@ -0,0 +1,461 @@
+{
+    var client = app.Services.GetService<IHttpClientFactory>().CreateClient();
+
+    var message = context.CreateProxyHttpRequest(new Uri("http://localhost:4301/inventory"));
+    var response = await client.SendAsync(message);
+    await context.CopyProxyHttpResponse(response);
+});
+
+app.MapGet("/sales/sku/{sku:required}", async context =>
+{
+    var sku = context.Request.RouteValues["sku"];
+    var client = app.Services.GetService<IHttpClientFactory>().CreateClient();
+    var message = context.CreateProxyHttpRequest(new Uri($"http://localhost:4302/sku/{sku}"));
+    var response = await client.SendAsync(message);
+    await context.CopyProxyHttpResponse(response);
+}).RequireAuthorization();
+
+app.Run();
+
+void AddAuthMiddleware(WebApplicationBuilder builder)
+{
+    // This method is all it takes to use Okta to automatically test incoming
+    // bearer tokens for authentication!
+    builder.Services.AddAuthentication(options =>
+    {
+        options.DefaultAuthenticateScheme = OktaDefaults.ApiAuthenticationScheme;
+        options.DefaultChallengeScheme = OktaDefaults.ApiAuthenticationScheme;
+        options.DefaultSignInScheme = OktaDefaults.ApiAuthenticationScheme;
+    })
+    .AddOktaWebApi(new OktaWebApiOptions()
+    {
+        OktaDomain = builder.Configuration["Okta:Domain"],
+    });
+    
+    builder.Services.AddAuthorization();
+}
+```
+
+Again, open up three separate terminals and navigate to the root directory for each of the .NET projects you created. Run all three services by executing `dotnet run` in each terminal.
+
+You'll have to provide a bearer token to the gateway service in order to access `/sales/sku/<sku>` but `/warehouse` will not require any authentication.
+
+To get an access token to access `/sales/sku/<sku>`, you'll need to execute the following request in Postman by using your own `client_id` and `client_secret` values:
+
+{% img blog/dotnet-microservices-federation/postman.png alt:"Request token in Postman" width:"600" %}{: .center-image }
+
+You should have a response with the `access_token` displayed. Copy the token and then issue the following request using Postman to replace the bearer token with your access token:
+
+{% img blog/dotnet-microservices-federation/successresponse.png alt:"Successful response in Postman" width:"600" %}{: .center-image }
+
+If implemented correctly, you should see a successful response!
+
+## Use Docker to run the solution
+
+In Visual Studio, right-click each project and click on **Add** > **Docker Support**. Then, choose Linux as the target OS. This will add a `Dockerfile` to each project that will be configured to build and run the .NET application with sensible defaults.
+
+Right-click the `ServiceGateway` project and choose **Add** > **Container orchestration support**. Choose **Docker Compose** and then **Linux** as the target OS. This step will add a `docker-compose.yml` file.
+
+![Visual Studio Docker tools](https://i.imgur.com/9ASEGwT.png)
```

Comment:
This image appears broken in this push...

Reply:
It should be fixed in the second push with the images. I replaced all the URLs